File: Cplusplus/Binary_Tree_Level_Order_Traversal_II.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int> > levelOrderBottom(TreeNode *root) {
        vector<vector<int>> result;
        if(root == NULL)
            return result;
        queue<TreeNode *> nodeQ;
        queue<int> levelQ;
        nodeQ.push(root);
        levelQ.push(1);
        int pre = 1;
        vector<int> levelVec;
        while(nodeQ.size())
        {
            TreeNode *node = nodeQ.front();
            int level = levelQ.front();
            nodeQ.pop();
            levelQ.pop();
            if(node->left != NULL)
            {
                nodeQ.push(node->left);
                levelQ.push(level+1);
            }
            if(node->right != NULL)
            {
                nodeQ.push(node->right);
                levelQ.push(level+1);
            }
            if(level == pre)
            {
                levelVec.push_back(node->val);
            }
            else
            {
                pre = level;
                result.push_back(levelVec);
                levelVec.clear();
                levelVec.push_back(node->val);
            }
        }
        result.push_back(levelVec);
        reverse(result.begin(),result.end());
        return result;
    }
};

File: python/Binary_Tree_Level_Order_Traversal_II.py
'''
Created on 2014.7.25

Leetcode : Binary Tree Level Order Traversal 

Given a binary tree, return the level order traversal of its nodes' values. (ie, from left to right, level by level).

For example:
Given binary tree {3,9,20,#,#,15,7},
    3
   / \
  9  20
    /  \
   15   7
return its level order traversal as:
[
  [3],
  [9,20],
  [15,7]
]

'''

# Definition for a  binary tree node

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of lists of integers
    def levelOrder(self, root):
        r = []
        if root == None:
            return r;
        queue = [root];
        level = [1]
        current = [];
        while len(queue) != 0:
            node =  queue[0];
            clevel = level[0];
            
            if node.left != None:
                queue.append(node.left);
                level.append(clevel+1);
            if node.right != None :
                queue.append(node.right);
                level.append(clevel+1);
            
            current.append(node.val);
            if len(level) == 1 or level[0] != level[1]:
                r.append(current[:]);
                current = []
            queue = queue[1:]
            level = level[1:]
        r.reverse()
        return r
test = Solution();
root = TreeNode(3)
node21 = TreeNode(9)
node22 = TreeNode(20)
node31 = TreeNode(15)
node32 = TreeNode(7)
node22.left = node31
node22.right = node32
root.left = node21
root.right = node22
print test.levelOrder(root)
                
            
                

